source/features/useful-not-found-page.tsx

```diff
@@ -29,71 +27,247 @@ async function checkAnchor(anchor: HTMLAnchorElement): Promise<void> {
+			linkToCommitHistory,
+		};
+	}
 }
 
-// If the resource exists in the default branch, link to it
-async function addDefaultBranchLink(bar: Element): Promise<void> {
-	const parts = getCleanPathname().split('/');
-	const branch = parts[3];
-	if (!branch) {
+async function getLinkToCommitHistoryOnDefaultBranch(): Promise<string | undefined> {
+	// Get the current branch
+	const url = new GitHubURL(location.href);
+	const currentBranch = url.branch;
+
+	if (!currentBranch) {
 		return;
 	}
 
 	const defaultBranch = await getDefaultBranch();
-	if (branch === defaultBranch) {
+	if (currentBranch === defaultBranch) {
 		return;
 	}
 
-	parts[3] = defaultBranch; // Change branch
-	const url = '/' + parts.join('/');
-	if (await is404(location.origin + url)) {
+	url.assign({
+		route: 'commits',
+		branch: defaultBranch,
+	})
+	// Check if that path exists
+	if (await is404(url.toString())) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			The {getType()} exists on the <a href={url}>default branch</a>.
-		</p>,
-	);
+	// If it does, return it
+	return url.toString();
 }
 
-function init(): false | void {
-	const parts = parseCurrentURL();
-	if (parts.length <= 1 || !select.exists('[alt*="This is not the web page you are looking for"]')) {
-		return false;
+async function whatHappenedToTheFile(): Promise<Record<string, string | undefined> | void> {
+	// First, check if the file exists on the default branch
+	const urlOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlOnDefaultBranch) {
+		return {
+			type: 'isOnDefaultBranch',
+			linkToLastVersion: urlOnDefaultBranch,
+		};
 	}
 
-	const bar = <h2 className="container mt-4 text-center"/>;
+	// Else, check if the file exists in commit history
+	const lastCommitIncludingFile = await getLastCommitForFile();
+	if (lastCommitIncludingFile) {
+		return lastCommitIncludingFile;
+	}
+
+	// Else, check if the file exists in the default branch's history
+	const linkToCommitHistoryOnDefaultBranch
+		= await getLinkToCommitHistoryOnDefaultBranch();
+	if (linkToCommitHistoryOnDefaultBranch) {
+		return {
+			type: 'wasOnDefaultBranch',
+			linkToCommitHistory: linkToCommitHistoryOnDefaultBranch,
+		};
+	}
+}
+
+async function showAdditionalInformation(bar: Element): Promise<void> {
+	// Check what happened
+	const eventThatHappened = await whatHappenedToTheFile();
+	if (!eventThatHappened) {
+		return;
+	}
+
+	// Parse that and do/show something
+	if (eventThatHappened.type === 'isOnDefaultBranch') {
+		bar.after(
+			<p className="container mt-4 text-center">
+				<a href={eventThatHappened.linkToLastVersion}>This file</a> only exists on the default branch.
+			</p>,
+		);
+		return;
+	}
+
+	if (eventThatHappened.type === 'wasOnDefaultBranch') {
+		bar.after(
+			<p className="container mt-4 text-center">
+				This file used to exist on the default branch - see <a href={eventThatHappened.linkToCommitHistory}>commit history</a>.
+			</p>,
+		);
+		return;
+	}
```

Comment:
This situation should just be covered by the deleted/renamed one, without mentioning it. Details in https://github.com/sindresorhus/refined-github/pull/4646#discussion_r684608765